File: 372 - Android/Ashman/app/src/main/java/com/benedetto/lars/ashman/entities/Entity.java
package com.benedetto.lars.ashman.entities;

import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;

import com.benedetto.lars.ashman.MazeView;
import com.benedetto.lars.ashman.enums.Direction;

import static com.benedetto.lars.ashman.enums.Direction.RIGHT;

public abstract class Entity {
    private static final int BLANK = 0;
    private static final int WALL_CAP = 1;
    private static final int WALL_CORNER = 2;
    private static final int WALL_INTERSECTION3 = 3;
    private static final int WALL_INTERSECTION4 = 4;
    private static final int WALL = 5;
    private static final int CAKE = 6;
    Direction directionActual;
    float X;
    float Y;
    public int Xindex;
    public int Yindex;
    private float velocity;
    public boolean INVINCIBLE;
    float tileSize;
    private int height;
    private int width;

    final MazeView mazeView;

    Entity(MazeView mazeView) {
        this.mazeView = mazeView;
        directionActual = RIGHT;
    }

    public abstract void draw(Canvas canvas, Paint paint, Matrix transformer);

    protected abstract int getWidth();

    protected abstract void hitWall();

    protected abstract void modify();

    public void setDirection(Direction direction) {
        directionActual = direction;
    }

    public void setVelocity(float velocity) {
        this.velocity = velocity;
    }

    public void move() {
        switch (directionActual) {
            case UP:
                Y -= velocity;
                if (Y < 0) {
                    Y = 0;
                    hitWall();
                } else if (isInWall(true, false, 0, 0)) {
                    Y += velocity;
                    hitWall();
                }
                break;
            case DOWN:
                Y += velocity;
                if ((Y + tileSize) > height || isInWall(true, true, 0, 0)) {
                    Y -= velocity;
                    hitWall();
                }
                break;
            case LEFT:
                X -= velocity;
                if (X < 0) {
                    X = 0;
                    hitWall();
                } else if (isInWall(true, false, 0, 0)) {
                    X += velocity;
                    hitWall();
                }
                break;
            case RIGHT:
                X += velocity;
                if ((X + tileSize) > width * 14 || isInWall(true, true, 0, 0)) {
                    X -= velocity;
                    hitWall();
                }
                break;
        }
        Xindex = (int) (X / tileSize);
        Yindex = (int) (Y / tileSize);
    }

    int getWallType(int Y, int X) {
        try {
            return Integer.valueOf(mazeView.maze[Y][X].split("")[1]);
        } catch (ArrayIndexOutOfBoundsException e) {
            return WALL;
        }
    }

    boolean isWall(int wallType) {
        switch (wallType) {
            case BLANK:
            case CAKE:
                return false;
            case WALL:
            case WALL_CAP:
            case WALL_CORNER:
            case WALL_INTERSECTION3:
            case WALL_INTERSECTION4:
                return true;
        }
        return true;
    }

    boolean isInWall(boolean modify, boolean bottomRight, int xOffset, int yOffset) {
        int buffer = getWidth();
        if (bottomRight) {
            Xindex = (int) ((X + buffer) / tileSize);
            Yindex = (int) ((Y + buffer) / tileSize);
        } else {
            Xindex = (int) (X / tileSize);
            Yindex = (int) (Y / tileSize);
        }
        int wallType = getWallType(Yindex + yOffset, Xindex + xOffset);
        if (wallType == CAKE && modify) modify();
        return isWall(wallType);
    }

    public void setTileSize(float tileSize) {
        this.tileSize = tileSize;
    }

    public void setWidth(int width) {
        this.width = width;
    }

    public void setHeight(int height) {
        this.height = height;
    }
}

File: 372 - Android/Ashman/app/src/main/java/com/benedetto/lars/ashman/entities/Entity.kt
package com.benedetto.lars.ashman.entities

import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint

import com.benedetto.lars.ashman.MazeView
import com.benedetto.lars.ashman.enums.Direction

import com.benedetto.lars.ashman.enums.Direction.RIGHT

abstract class Entity internal constructor(internal val mazeView: MazeView) {
    internal var directionActual: Direction
    internal var x: Float = 0.toFloat()
    internal var y: Float = 0.toFloat()
    var xIndex: Int = 0
    var yIndex: Int = 0
    private var velocity: Float = 0.toFloat()
    var isInvincible: Boolean = false
    internal var tileSize: Float = 0.toFloat()
    private var height: Int = 0
    private var width: Int = 0

    init {
        directionActual = RIGHT
    }

    abstract fun draw(canvas: Canvas, paint: Paint, transformer: Matrix)

    protected abstract fun getWidth(): Int

    protected abstract fun hitWall()

    protected abstract fun modify()

    fun setDirection(direction: Direction) {
        directionActual = direction
    }

    fun setVelocity(velocity: Float) {
        this.velocity = velocity
    }

    fun move() {
        when (directionActual) {
            Direction.UP -> {
                y -= velocity
                if (y < 0) {
                    y = 0f
                    hitWall()
                } else if (isInWall(true, false, 0, 0)) {
                    y += velocity
                    hitWall()
                }
            }
            Direction.DOWN -> {
                y += velocity
                if (y + tileSize > height || isInWall(true, true, 0, 0)) {
                    y -= velocity
                    hitWall()
                }
            }
            Direction.LEFT -> {
                x -= velocity
                if (x < 0) {
                    x = 0f
                    hitWall()
                } else if (isInWall(true, false, 0, 0)) {
                    x += velocity
                    hitWall()
                }
            }
            RIGHT -> {
                x += velocity
                if (x + tileSize > width * 14 || isInWall(true, true, 0, 0)) {
                    x -= velocity
                    hitWall()
                }
            }
        }
        xIndex = (x / tileSize).toInt()
        yIndex = (y / tileSize).toInt()
    }

    internal fun getWallType(Y: Int, X: Int): Int {
        return try {
            Integer.valueOf(mazeView.maze[Y][X].split("".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1])
        } catch (e: ArrayIndexOutOfBoundsException) {
            WALL
        }

    }

    internal fun isWall(wallType: Int): Boolean {
        return when (wallType) {
            BLANK, CAKE -> false
            else -> return true
        }
    }

    internal fun isInWall(modify: Boolean, bottomRight: Boolean, xOffset: Int, yOffset: Int): Boolean {
        val buffer = getWidth()
        if (bottomRight) {
            xIndex = ((x + buffer) / tileSize).toInt()
            yIndex = ((y + buffer) / tileSize).toInt()
        } else {
            xIndex = (x / tileSize).toInt()
            yIndex = (y / tileSize).toInt()
        }
        val wallType = getWallType(yIndex + yOffset, xIndex + xOffset)
        if (wallType == CAKE && modify) modify()
        return isWall(wallType)
    }

    open fun setTileSize(tileSize: Float) {
        this.tileSize = tileSize
    }

    fun setWidth(width: Int) {
        this.width = width
    }

    fun setHeight(height: Int) {
        this.height = height
    }

    companion object {
        private const val BLANK = 0
        private const val WALL = 5
        private const val CAKE = 6
    }
}
